src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,463 @@
+        this instance.
+
+        NOTE: the necessary dimensions for the jacobian can be extracted from the dimensions
+        of the jacobian of passed Ad_array instances.
+        """
+        pass
+
+
+class AbstractJacFunction(AbstractFunction):
+    """'Half'-abstract base class, for which a direct evaluation of the passed callable is
+    implemented. Only the Jacobian has to be concretized / approximated."""
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        """
+        Evaluates the passed callable directly using values of Ad_array instances
+        passed as arguments.
+
+        :param args: tuple of :class:`~porepy.numerics.ad.forward_mode.Ad_array`
+        :type args: tuple
+        :return: results from callable packed into an array.
+        :rtype: numpy.array
+        """
+        # get values of argument Ad_arrays.
+        vals = (arg.val for arg in args)
+
+        # if the callable is flagged as conform for vector operations, feed vectors
+        if self.is_vector_func:
+
+            return self.func(*vals)
+        else:
+            # if not vector-conform, feed element-wise
+
+            # TODO this displays some special behavior when val-arrays have different lengths:
+            # it returns None-like things for every iteration more then shortest length
+            # These Nones are ignored for some reason by the function call, as well as by the
+            # array constructor.
+            # If a mortar var and a subdomain var are given as args,
+            # then the lengths will be different for example.
+            return np.array([self.func(*vals_i) for vals_i in zip(*vals)])
+
+
+### CONCRETE IMPLEMENTATIONS ------------------------------------------------------------------
+
+
+class Function(AbstractFunction):
+    """Ad representation of an analytically given function,
+    where it is expected that feeding the callable with Ad_arrays directly will give
+    give the proper result.
+
+    Here the values AND the Jacobian are obtained exactly by standard rules of differentiation.
+
+    The intended use is as a wrapper for operations on pp.ad.Ad_array objects,
+    in forms which are not directly or easily expressed by the rest of the Ad
+    framework.
+
+    NOTE: this is a special case where `get_values` and `get_jacobian` are formally implemented
+    but never used by the AD framework
+    (due to the assumed, direct evaluation of result using the callable).
+    """
+
+    def __init__(
+        self, func: Callable, name: str, is_vector_func: Optional[bool] = True
+    ):
+        super().__init__(func, name, is_vector_func, True)
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.val
+
+    def get_jacobian(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.jac
```

Comment:
I don't like this signature at all. If `self.func` is costly (and somtimes it will be), we will have twice the computational cost for no clear reason.

It seems to me the need to split this comes from forcing `pp.ad.Function` to be on the same inheritance level at the wrapping of an Ad_array with known value and Jacobian. It is not clear this is the right design choice (my instinct is it is wrong).

Reply:
The `get_values` and `get_jacobian` methods of this re-designed `pp.ad.Function` are never called by the framework
(see file `operators.py` line 422)

The idea to put them together is motivated by the concept of a function (and the possibility to unify it) and to save code.
Otherwise we would quite some duplicate code in `Function` and `AbstractFunction`. Also in the `Operator` class, we would have to different "evaluate" operations.

What we can fall back on is the unfinished "approximate" operation, which was not completely implemented by Jakub.
Then we would keep the concept of an analytic `Function` and introduce the concept of approximating `AbstractFunction`.
But since they are so similar, it is worth a shot to unify them somehow. My first attempt was this hierarchy and the argument `is_adarray_func`.